cranelift/codegen/src/isa/pulley_shared/lower.isle

```diff
@@ -226,19 +226,32 @@
 ;;;; Rules for `load` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (rule (lower (has_type ty (load flags addr (offset32 offset))))
-      (pulley_load (Amode.RegOffset addr (i32_as_i64 offset))
-                   ty
-                   flags
-                   (ExtKind.Zero)))
+    (let ((le Reg (pulley_load (Amode.RegOffset addr (i32_as_i64 offset))
+                               ty
+                               flags
+                               (ExtKind.Zero))))
+      (xswap_if_be le ty flags)))
+
 
 ;;;; Rules for `store` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
+(decl littleendian () MemFlags)
+(extern extractor littleendian littleendian)
+
+(decl bigendian () MemFlags)
+(extern extractor bigendian bigendian)
+
 (rule (lower (store flags src @ (value_type ty) addr (offset32 offset)))
       (side_effect (pulley_store (Amode.RegOffset addr (i32_as_i64 offset))
-                                 src
+                                 (xswap_if_be src ty flags)
                                  ty
                                  flags)))
 
+(decl xswap_if_be (XReg Type MemFlags) XReg)
+(rule 0 (xswap_if_be val _ty (littleendian)) val)
+(rule 1 (xswap_if_be val $I32 (bigendian)) (pulley_bswap32 val))
+(rule 1 (xswap_if_be val $I64 (bigendian)) (pulley_bswap64 val))
```

Comment:
... which we can then use like

```suggestion
(decl xswap_if_be (XReg Type MemFlags) XReg)
(rule 0 (xswap_if_be val _ty flags)
      (if-let (endian flags) (Endian.Little))       
      val)
(rule 1 (xswap_if_be val $I32 flags)
      (if-let (endian flags) (Endian.Big))
      (pulley_bswap32 val))
(rule 1 (xswap_if_be val $I64 flags)
      (if-let (endian flags) (Endian.Big))
      (pulley_bswap64 val))
```

and I think we could remove the explicit priorities at that point because isle's overlap checking would be able to see that the rules do not overlap, which would also result in better isle-generated rust code.